walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,84 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) { 
+    // check edge cases before calling helper functions
+    if (request.getDuration() == 0 || request.getDuration() > 1440 ) { 
+        return new ArrayList<>();
+    }
+
+    List<TimeRange> unavailableTimeRanges = getUnavailableTimeRanges(events, request);
+    List<TimeRange> availableTimeRanges = getAvailableTimeRanges(unavailableTimeRanges, request.getDuration());
+
+    return availableTimeRanges;
+  }
+
+  /**
+  * Returns the time ranges for which the meeting request cannot coincide with. 
+  * A time range isn't available if there is an event happening and some people in the
+  * event are required to be in the meeting.
+  * If there are no unavailable slots, this function returns an empty arraylist.
+  *
+  * @param events the existing events with attendees that already planned
+  * @param request the meeting request containing the duration and attendees needed
+  * @return an arraylist containing TimeRanges that aren't available for the meeting, sorted by start time
+  */
+  private List<TimeRange> getUnavailableTimeRanges(Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
+
+    for (Event e : events) {
+        // if there are attendees in common, that time range isn't available 
+        if (!Collections.disjoint(e.getAttendees(), request.getAttendees())) {
+            unavailableTimeRanges.add(e.getWhen());
+        }
+    }
+
+    // sort the time ranges in chronological order
+    Collections.sort(unavailableTimeRanges, TimeRange.ORDER_BY_START);
+
+    return unavailableTimeRanges;
+  }
+
+  /**
+  * Returns the time ranges for which the meeting request can happen in. 
+  * This returns the "inverse" of the unavailable timeslots.
```

Comment:
It's not exactly the inverse, right? Maybe add a note that it only adds time slots with a duration greater than the required duration